### DISCUSSION TITLE
Veb & templates

### DISCUSSION THREAD

**Discussion by kylepritchard** [vlang/v#21868] [Questions and Answers]
I am currently experimenting with veb and the templating system but have encountered an issue which I want to ask about before submitting a bug. I am trying to include a component using @include directive but I can not get an error with the path. Is it possible to use the '../' to move up one level in the file path or do all includes need to be in same direction or child directory of the template calling the @include directive?

```
@for t in app.todos
<!-- todo component -->
@include "../todo.html"
<!-- end -->
@end
```
 results in below 

``` bash
src/templates/index.html:-6:1: error: Reading file ../todo from path: /home/kyle/Programming/v/htmx/src/templates/../todo.html failed
    1 | <html>
Details: Failed to @include '../todo'
```

**Comment by changrui** [comment]
Put your HTML files to `./templates` of project directory please.

**Reply by kylepritchard** [reply]
I have done that but the templates will not render. The problem comes when using a relative path. If a template is called from a nested child folder it will not include a template file from the parent folder or a sibling folder to the parent. 

I am looking into it and I believe there may be an issue with the _os.join_path_single_ (mentioned in [#21751](https://github.com/vlang/v/issues/21751)) as it will not take into account the '../' backtick as part of the path and instead it seems to be included as part of the absolute path. It could also be an issue with _os.real_path_

I am still looking into it more to see if I am right or wrong and see if i can make a PR to fix it. 

This could also just be an error on my system (Ubuntu 24.04)

**Comment by kylepritchard** [comment]
I finally had time to look more at this and the fault was at my end, kind of. When tracing the template rendering, the templates are rendered relative to the first called file which works if rendering an entire page which calls all the @include's. This works well in a traditional web page which is a complete reload of the page. The current trending pattern is to call html partials and swap portions of the page using frameworks such as HTMX and Unpoly. When rendering a complete page the rendering will all be done with the base path being the directory of the root html file, however, if a partial of that page is rendered then the basepath will be the directory containing the partial html file. If both the main page and partial share a component then there will be issues finding the components.

As an example

![tree](https://github.com/user-attachments/assets/33a4b89f-bd23-41ae-b0a2-3f4a65cec55d)

In the above folder the index.html file is basically a layout and will @include main.html -> parent.html -> sections/worklist.html -> components/todo.html in that order. This is easy to do as the basepath will be the templates folder so all includes are relative to the root folder

However, if sections/worklist was to be rendered itself and updated into the DOM then to include components/todo.html  the @include would need to look like ```@include '..\components\todo' ```. This now unfortunately doesn't work with the call from index.html as it would now have an absolute path one directory up from the templates folder which of course doesnt work.

![error](https://github.com/user-attachments/assets/d27defc9-3010-419b-8660-7b6733592bfa)

The solution?

Currently the only way to do this is to ensure all child elements on a page/template are filed in such a way that they are in same directory as the calling template or deeper. This can become difficult to manage when developing a complicated app with many components.

I am wondering if there is a better way of enabling calling templates relative to the templates folder? I had a few ideas which i could try to implement and make a PR if it is of interest or if people agree. They should be easy to implement and will not result in any breaking changes

``` @include! 'path/to/template' ``` - Adding the bang to @include would call a small bit of logic which walks up the tree to find the root folder and then the template can be found from there

``` @include '~/path/to/folder' ``` - This would be similar but using a tilda to start the path would tell the program to walk up the tree to the root directory and then from there 

Not sure what peoples thoughts are on these. I would love to try and rebuild a complicated web based app in V but this part is holding me back as I know from past experience that organising the templates folder can become a challenge when sharing components and the only way to build a full app is by reusing small components.
